### PR TITLE
Ex6[p] diagonal preconditioner for PA

### PR DESCRIPTION
**Pull request by stefanhenneking**
Adding diagonal preconditioner for PA to the AMR example ex6[p].

Note: depends on #1618 

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge| 
 | --- | --- | --- | --- | --- | --- | --- | 
| [#1632](https://github.com/mfem/mfem/pull/1632) | @stefanhenneking | @tzanio | @barker29 + @jakubcerveny + @tzanio | 08/16/20 | 08/21/20 | 09/01/20 | |

### CODE REVIEW

File: examples/ex6p.cpp [hunk 1]
```diff
@@ -220,17 +224,26 @@ int main(int argc, char *argv[])
...
+      Solver *M = NULL;
+      if (pa)
+      {
+         M = new OperatorJacobiSmoother(a, ess_tdof_list);
+      }
+      else
+      {
+         HypreBoomerAMG *amg = new HypreBoomerAMG;
+         amg->SetPrintLevel(0);
+         M = amg;
```
Comment (jakubcerveny) [inline examples/ex6p.cpp:237]: The `amg` variable is now unneccessary here...

File: examples/ex6p.cpp [hunk 2]
```diff
@@ -129,7 +129,11 @@ int main(int argc, char *argv[])
    //    the Laplace problem -\Delta u = 1. We don't assemble the discrete
    //    problem yet, this will be done in the main loop.
    ParBilinearForm a(&fespace);
-   if (pa) { a.SetAssemblyLevel(AssemblyLevel::PARTIAL); }
+   if (pa)
+   {
+      a.SetAssemblyLevel(AssemblyLevel::PARTIAL);
+      a.SetDiagonalPolicy(mfem::Operator::DIAG_ONE);
```
Comment (jakubcerveny) [inline examples/ex6p.cpp:135]: Do we need the explicit `mfem::` namespace?

Comment (stefanhenneking) [reply examples/ex6p.cpp]: You are right, we don't need it; I removed it here.

File: examples/ex6p.cpp (hunk 1 above)
Comment (stefanhenneking) [reply examples/ex6p.cpp:237]: To set the print level of the `HypreBoomerAMG` solver via `SetPrintLevel`, I would then have to use
```
M = new HypreBoomerAMG;
static_cast<HypreBoomerAMG*>(M)->SetPrintLevel(0);
```
Personally, I think using variable `amg` makes it easier to read but I can change it if casting this way is preferred.

Comment (jakubcerveny) [reply examples/ex6p.cpp:237]: Sorry, I overlooked the fact that the pointers are not the same type. Please disregard.

**Comment by tzanio** [comment]
@jakubcerveny, if all your comments have been addressed can you approve?

**Comment by v-dobrev** [comment]
In parallel, I'm not seeing as much reduction of the iterations as in serial. For example the sample run
```console
mpirun -np 4 ex6p -pa -d cuda -no-vis
```
ends with
```diff
 AMR iteration 26
 Number of unknowns: 123036
-   Iteration :   0  (B r, r) = 7.30948e-06 ...
-   Iteration : 314  (B r, r) = 7.23743e-18
-Average reduction factor = 0.95694
+   Iteration :   0  (B r, r) = 2.57067e-06 ...
+   Iteration : 295  (B r, r) = 2.30715e-18
+Average reduction factor = 0.954073
 Reached the maximum number of dofs. Stop.
```
On the other hand, the serial sample run
```console
ex6 -pa -d cuda -no-vis
```
ends with
```diff
 AMR iteration 19
 Number of unknowns: 56668
-   Iteration :   0  (B r, r) = 3.11646e-05 ...
-   Iteration : 609  (B r, r) = 2.50296e-17
-Average reduction factor = 0.977394
+   Iteration :   0  (B r, r) = 8.24396e-06 ...
+   Iteration : 339  (B r, r) = 6.88238e-18
+Average reduction factor = 0.95981
 Reached the maximum number of dofs. Stop.
```
Any thoughts on why there is such a difference? Could this be an indication of an issue with the parallel version?

(On the other hand, the parallel version was already doing just half of the iterations of the serial version. The two versions use different error estimators but it is not clear if that can account for the differences.)

**Comment by v-dobrev** [comment]
Also, @jakubcerveny said:

> In ex6p there's about 1/2 iterations

but the internal autotest (that's where the above diffs are taken from) does not agree with that.

**Comment by tzanio** [comment]
> This is very nice 👍, I'm getting 1/3 iterations in ex6 and about 1/2 total time. In ex6p there's about 1/2 iterations, but as a side note, we should do something with the extremely slow `L2ZienkiewiczZhuEstimator` (contains a global assembly/solve), it now takes ~90% of the time in ex6p...

@jakubcerveny can you post the exact runs where you see this?

**Comment by jakubcerveny** [comment]
Sorry for the late response, I was on vacation. I was running something like this:
```
mpirun -np 6 ex6p -m ../data/fichera.mesh -no-vis -o 4 -pa
```
I ran with the new `OperatorJacobiSmoother` and a version without the preconditioner (`M == NULL`). This is where I saw ~113 iterations vs. ~229.

**Comment by tzanio** [comment]
Extending the merge deadline to address @v-dobrev's comments...

**Comment by v-dobrev** [comment]
Thanks @jakubcerveny. You used higher-order elements, so that explains the better convergence rates you see with diagonal preconditioning.

Another test that @psocratis ran was to try the error estimator from `ex6` (i.e. `ZienkiewiczZhuEstimator`) in the parallel version, `ex6p`, and with that he observed the same results as in `ex6`, i.e. improved convergence rates. Is that correct, @psocratis?

**Comment by psocratis** [comment]
That is correct.

**Comment by v-dobrev** [comment]
> That is correct.

Okay, then the parallel implementation must be fine and the error estimator (interestingly) is the reason for this behavior.

So, this PR should be ready for merging.

**Comment by tzanio** [comment]
Nice work @stefanhenneking !

**Pull request merged**